source/extensions/common/wasm/wasm.cc

```diff
@@ -654,6 +654,7 @@ createWasmInternal(const VmConfig& vm_config, PluginSharedPtr plugin, Stats::Sco
             }
             if (!root_context) {
               wasm->wasm()->start(plugin);
+              wasm->wasm()->clearRootContexts();
```

Comment:
does clearning root context like this have wide implecations?
How will it affect other plugins?

Reply:
+1

Reply:
This is the base vm.  So we do need to talk about the lifecycle of the base Wasm vm.   It is not a real VM in the sense that it will never receive any events.  Moreover since it is in the global cache it has been divorced in some sense from the creating thread which in any case is often the main thread.   The issue were were seeing is that it is managed by smart pointers, std::shared_ptr and std::weak_ptr (in the cache).  Consequently it ends up being deleted on some random Worker thread which is never the thread that it is created on.  The Wasm VMs are single threaded and moreover the wrapper function accesses thread_local variables from the creating thread.  The net net of this is that running the onDone call in response to the destructor on some random thread is a bad idea.  Now we could transition the destructor back to the creating thread (assuming that it has a Dispatcher) but again, why? the only reason we called start() on this VM was so that it would parse the vm_configuration and possibly toss an error during the XDS configuration update so that we can fail the configuration updated.  It makes little sense to have the full lifecycle.   However, perhaps this is a situation where we should re-vivify validate_configuration() instead of reusing start() for the base VM.  Something to think about for the ABI lifecycle section.   In the short term this is more like running kill -9 on the base Wasm VM.  It has consequences, but not likely ones which will be adverse.

Reply:
I am wondering why is base wasm deleted on a random worker thread? It is not obvious to me how would that happen.

Also I thought the purpose of base VM is to save some loading time/cpu by cloning it to thread local Wasm, no? Does it really help configuration validation, considering the same old reason that XDS does not support async response?

Reply:
We added a remote fetch cache and now require that there be a cache hit so that we can create, initialize and call into the VM to validate the VM configuration inline of the xDS update so that we can return the xDS status synchronously.  This is done sychronously via the onStart call into the VM on a Base Wasm object.   This object is use for this purpose and also as the template for cloning the per-thread VMs.  This VM is stored in a global hash table of weak_ptr.  This VM is also stored in a shared_ptr attached to each thread/Worker VM  so that when the last of these thread/Worker VMs is deleted we can cleanup the corresponding Base Wasm VM.  Consequently it is deleted on some random thread/Worker when the last thread/Worker is deleted when the last thread/Worker Listener is deleted.

Reply:
This process is our work-around for the lack of async response support in xDS>